File: 프로그래머스/2/12905. 가장 큰 정사각형 찾기/가장 큰 정사각형 찾기.py
def solution(board):
    answer = 0
    
    row, col = len(board), len(board[0])
    n = max(row, col)
    
    dp = [[0] * (n + 1) for _ in range(n + 1)]

    for i in range(1, len(board) + 1):
        for j in range(1, len(board[0]) + 1):
            if board[i-1][j-1] != 0:
                dp[i][j] = min(dp[i][j-1], dp[i-1][j], dp[i-1][j-1]) + 1
                answer = max(answer, dp[i][j])
    
    return answer ** 2

File: 프로그래머스/2/12905. 가장 큰 정사각형 찾기/가장 큰 정사각형 찾기.java
class Solution {
    public int solution(int[][] board) {
        int answer = 0;

        int row = board.length;
        int col = board[0].length;
        int n = Math.max(row, col);

        int[][] dp = new int[n + 1][n + 1];

        for (int i = 1; i <= row; i++) {
            for (int j = 1; j <= col; j++) {
                if (board[i - 1][j - 1] != 0) {
                    dp[i][j] = Math.min(
                        Math.min(dp[i][j - 1], dp[i - 1][j]),
                        dp[i - 1][j - 1]
                    );
                    answer = Math.max(answer, ++dp[i][j]);
                }
            }
        }

        return answer * answer;
    }
}